libs/datamodel/core/src/transform/ast_to_dml/lift.rs

```diff
@@ -212,204 +178,181 @@ impl<'a> LiftAstToDml<'a> {
-                        )),
-                    }
-                } else {
-                    Ok((dml::FieldType::Scalar(scalar_type, type_alias, None), vec![]))
-                }
-            } else {
-                Ok((dml::FieldType::Scalar(scalar_type, type_alias, None), vec![]))
-            }
-        } else if self.db.ast().find_model(type_name).is_some() {
-            Ok((dml::FieldType::Relation(dml::RelationInfo::new(type_name)), vec![]))
-        } else if self.db.get_enum(type_name).is_some() {
-            Ok((dml::FieldType::Enum(type_name.clone()), vec![]))
-        } else {
-            self.resolve_custom_type(ast_field, checked_types)
+fn lift_scalar_field_type(
+    ast_field: &ast::Field,
+    scalar_field_type: &ScalarFieldType,
+    collected_attributes: &mut Vec<ast::Attribute>,
+    db: &ParserDatabase<'_>,
+    diagnostics: &mut Diagnostics,
+) -> dml::FieldType {
+    match scalar_field_type {
+        ScalarFieldType::Enum(enum_id) => {
+            let enum_name = &db.ast()[*enum_id].as_enum().unwrap().name.name;
+            dml::FieldType::Enum(enum_name.to_owned())
+        }
+        ScalarFieldType::Unsupported => lift_unsupported_field_type(
+            ast_field,
+            ast_field.field_type.as_unsupported().unwrap().0,
+            db.datasource(),
+            diagnostics,
+        ),
+        ScalarFieldType::Alias(top_id) => {
+            let alias = db.ast()[*top_id].as_type_alias().unwrap();
+            collected_attributes.extend(alias.attributes.iter().cloned());
+            lift_scalar_field_type(
+                alias,
+                db.alias_scalar_field_type(top_id),
+                collected_attributes,
+                db,
+                diagnostics,
+            )
+        }
+        ScalarFieldType::BuiltInScalar => {
+            let scalar_type: dml::ScalarType = ast_field.field_type.unwrap_supported().name.parse().unwrap();
+            let native_type = db
+                .datasource()
+                .and_then(|datasource| lift_native_type(ast_field, &scalar_type, datasource, diagnostics));
+            dml::FieldType::Scalar(scalar_type, None, native_type)
         }
     }
+}
 
-    fn resolve_custom_type(
-        &self,
-        ast_field: &ast::Field,
-        checked_types: &mut Vec<String>,
-    ) -> Result<(dml::FieldType, Vec<ast::Attribute>), DatamodelError> {
-        let field_type = ast_field.field_type.unwrap_supported();
-        let type_name = &field_type.name;
-
-        if checked_types.iter().any(|x| x == type_name) {
-            // Recursive type.
-            return Err(DatamodelError::new_validation_error(
-                &format!(
-                    "Recursive type definitions are not allowed. Recursive path was: {} -> {}.",
-                    checked_types.join(" -> "),
-                    type_name
-                ),
-                field_type.span,
-            ));
-        }
+fn lift_native_type(
```

Comment:
This one next?

Reply:
Hehe I'm a bit afraid to touch this one, but it doesn't look that scary anymore on its own. You can have a shot if you want. We'll get to it at some point.